backend/zkppSalt.js

```diff
@@ -1,27 +1,69 @@
+//      registration or login session and the impact of changing them is limited
+//      to ongoing login or registration sessions (note that this means the login
+//      process itself, not the 'session' created after login, which is
+//      unaffected). This means that in theory the impact of rotating these
+//      values is rather low. The only reason to use a stable value for these
+//      parameters is to ensure that users are as little inconvenienced as
+//      possible if the server is restarted or if they are handed over to a
+//      different machine (e.g., under a load balancer).
+// When migrating to a different server, note the following:
+//   1. For internal migrations (i.e., the server administrator doesn't change):
+//      When transfering data over to a new instance, ensure that the key
+//      `_private_immutable_zkpp_ikm` is moved over and that it's the same as
+//      for the old instance.
+//   2. For other migrations (when the server administrator changes):
+//      a. Instances could implement a method to transfer salt records from one
+//         server to the other, either in an encrypted or unencrypted form.
+//         Since the actual encryption key used for `_private_rid_` entries is
+//         different for each contract ID, either key is possible. However, this
+//         is not implemented.
+//      b. Alternatively, migration can be done without migrating password salt
+//         records. This requires user interaction to create new salt records
+//         on the new server.
+const IkmPromise = sbp('chelonia/db/get', '_private_immutable_zkpp_ikm').then((IKM) => {
+  if (!IKM) {
+    const secret = randomBytes(33).toString('base64')
+    return sbp('chelonia/db/set', '_private_immutable_zkpp_ikm', secret).then(() => {
+      return secret
+    })
+  }
+  return IKM
+})
 
-// TODO HARDCODED VALUES
-// These values will eventually come from server the configuration
-const recordPepper = 'pepper' // TODO: get rid of `recordPepper`, just use `private/rid/${contractID}`
 // used to encrypt salts in database
-const recordMasterKey = 'masterKey' // TODO: store in file that's got root privs (after dropping privs)
+const recordSecret = IkmPromise.then(IKM => Buffer.from(hashStringArray('private/recordSecret', IKM)).toString('base64'))
 // corresponds to the key for the keyed Hash function in "Log in / session establishment"
-const challengeSecret = 'secret' // TODO: generate randomly and store in DB under private prefix
+const challengeSecret = IkmPromise.then(IKM => Buffer.from(hashStringArray('private/challengeSecret', IKM)).toString('base64'))
 // corresponds to a component of s in Step 3 of "Salt registration"
-const registrationSecret = 'secret' // TODO: generate randomly and store in DB under private prefix
+const registrationSecret = IkmPromise.then(IKM => Buffer.from(hashStringArray('private/registrationSecret', IKM)).toString('base64'))
```

Comment:
These values can be set in a `initZkpp()` function that gets called from `database.js` (see bottom of that file)